nheko: add module

gvolpe opened the pull request:
### Description

Add module for the [nheko](https://github.com/Nheko-Reborn/nheko) program.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

sumnerevans commented on tests/default.nix:
```diff
@@ -95,6 +95,7 @@ import nmt {
     ./modules/programs/newsboat
     ./modules/programs/nix-index
     ./modules/programs/nnn
+    ./modules/programs/nheko
```
"h" comes before "i" so this should be between newsboat and nix-index

gvolpe replied:
damn, a hiccup 🥲

sumnerevans commented on modules/programs/nheko.nix:
```diff
@@ -0,0 +1,83 @@
+let
+  cfg = config.programs.nheko;
+
+  iniType = with types;
+    let
+      primitiveType = either str (either bool (either int float));
+      multipleType = either primitiveType (listOf primitiveType);
+      sectionType = attrsOf multipleType;
+      supersectionType = attrsOf (either multipleType sectionType);
+    in attrsOf supersectionType;
```
This looks like a pretty standard INI file... we should be using `pkgs.formats.ini`

(See `modules/services/pantalaimon.nix` as an example)

gvolpe replied:
Oh nice, wasn't aware of that, I'll check it out!

sumnerevans commented on modules/programs/nheko.nix:
```diff
@@ -0,0 +1,83 @@
+        Attribute set of Nheko preferences (converted to INI file).
+
+        For now, it is recommended to run nheko and sign-in before filling in the configuration
+        settings in this module, as nheko writes the access token to $HOME/.config/nheko/nheko.conf
```
```suggestion
        settings in this module, as nheko writes the access token to <filename>$HOME/.config/nheko/nheko.conf</filename>
```

gvolpe commented:
@sumnerevans thanks for the review! I believe I addressed all your points and even fixed the tests, as the `nmt.script` line was in the wrong place in one of the tests 😅

gvolpe commented:
`nheko` seems to be marked as broken for MacOS, anything I should do here?

berbiche commented on modules/programs/nheko.nix:
```diff
@@ -0,0 +1,82 @@
+          };
+        }
+      '';
+      description = ''
+        Attribute set of Nheko preferences (converted to INI file).
+
+        For now, it is recommended to run nheko and sign-in before filling in the configuration
+        settings in this module, as nheko writes the access token to <filename>$XDG_CONFIG_HOME/nheko/nheko.conf</filename>
+        the first time we sign in, and we need that data into these settings for the correct
+        functionality of the application.
+
+        This a temporary inconvenience, however, as nheko has plans to move the authentication
+        stuff into the local database they currently use. Once this happens, this will no
+        longer be an issue.
+      '';
```
```suggestion
      description = ''
        Attribute set of Nheko preferences (converted to INI file).
        </para><para>
        For now, it is recommended to run nheko and sign-in before filling in the configuration
        settings in this module, as nheko writes the access token to <filename>$XDG_CONFIG_HOME/nheko/nheko.conf</filename>
        the first time we sign in, and we need that data into these settings for the correct
        functionality of the application.
        </para><para>
        This a temporary inconvenience, however, as nheko has plans to move the authentication
        stuff into the local database they currently use. Once this happens, this will no
        longer be an issue.
      '';
```

gvolpe replied:
Aren't those tags in reverse order? I see a closing tag first, then an opening tag, not sure if that's correct.

gvolpe replied:
Added them anyway, as suggested.

berbiche commented on modules/programs/nheko.nix:
```diff
@@ -0,0 +1,82 @@
+      text = ''
+        ; Generated by Home Manager.
+
+        ${generators.toINI { } (attrsKeyToSnakeCase cfg.config)}
```
```nix
generators.toINI {
  mkKeyValue = k: v: mkKeyValueDefault {} "=" (camelCaseToSnakeCase k) v;
} cfg.config
```

gvolpe replied:
Could you explain why this longer version would be preferred?

gvolpe replied:
Oh I guess we'd get rid of the `attrsKeyToSnakeCase` intermediate function, making it shorter in the end, is that correct?

gvolpe replied:
I think I got it. It's working now as suggested.

berbiche commented on tests/modules/programs/nheko/nheko-example-settings.nix:
```diff
@@ -0,0 +1,80 @@
+{ pkgs, ... }:
+
+{
+  config = {
```
```suggestion
{ ... }:

{
```

berbiche commented on tests/default.nix:
```diff
@@ -93,6 +93,7 @@ import nmt {
     ./modules/programs/ne
     ./modules/programs/neomutt
     ./modules/programs/newsboat
+    ./modules/programs/nheko
```
> `nheko` seems to be marked as broken for MacOS, anything I should do here?

The package needs to be stubbed in the tests.

gvolpe replied:
Ah that's what those stubs were for! I had them before but removed them because the tests were passing anyway, but of course, I'm on NixOS not on Mac 😄

berbiche commented on tests/modules/programs/nheko/nheko-empty-settings.nix:
```diff
@@ -0,0 +1,11 @@
+
+{
+  config = {
+    programs.nheko.enable = true;
```
```suggestion
    programs.nheko.enable = true;
    
    test.stubs.nheko = { };
```

berbiche commented on tests/modules/programs/nheko/nheko-example-settings.nix:
```diff
@@ -0,0 +1,80 @@
+        };
+      };
+
+    };
+
+    nmt.script = ''
```
```suggestion
    };
    
    test.stubs.nheko = { };

    nmt.script = ''
```

berbiche reviewed (changes requested):
Thanks for the contribution!

gvolpe commented:
Thanks for your review, @berbiche ! Left one or two questions, but I addressed all your points.

rycee commented on modules/programs/nheko.nix:
```diff
@@ -0,0 +1,87 @@
+  else
+    config.xdg.configHome;
+
+  camelCaseToSnakeCase =
+    replaceStrings upperChars (map (s: "_${s}") lowerChars);
```
We typically don't bother doing this for RFC-42 type `settings` options. That is, having 

    general.disable_certificate_validation = false

is OK.

gvolpe replied:
Can we have it this way then? Or RFC-42 prohibits this? I really dislike snake case on nix settings by default 😅

rycee replied:
Typically, the description of `settings` options contain an upstream link to where the user can find information about available program options. That upsteam documentation would show the snake-case names so it may confuse a user when they see non-snake-case names here.

But, I have no strong feeling about it so you can keep the `camelCaseToSnakeCase` if you want. As long as it is unambiguous, i.e., nheko doesn't have any option that contains a uppercase character.

gvolpe replied:
I'm afraid all these settings are very much under-documented, I had to dig them up, so no documentation link...

rycee commented on modules/lib/maintainers.nix:
```diff
@@ -305,4 +305,10 @@
     github = "lukasngl";
     githubId = 69244516;
   };
+  gvolpe = {
+    name = "Gabriel Volpe";
+    email = "volpegabriel@gmail.com";
+    github = "gvolpe";
+    githubId = 443978;
+  };
```
Since you are already in Nixpkgs' `maintainers/maintainer-list.nix` file you don't need to be here:

```suggestion
```

rycee commented on modules/programs/nheko.nix:
```diff
@@ -0,0 +1,87 @@
+
+  inherit (generators) mkKeyValueDefault toINI;
+in {
+  meta.maintainers = [ hm.maintainers.gvolpe ];
```
```suggestion
  meta.maintainers = [ maintainers.gvolpe ];
```

gvolpe commented:
@rycee thanks for your review, addressed all your points 👍🏽

gvolpe commented:
Hi folks, appreciate if anyone could hit the "Approve and run" button for the CI jobs, and perhaps give it a final review. I believe I addressed all your points, thank you @rycee @berbiche @sumnerevans  :)

rycee commented:
Nice, merged to master now 🙂
